sno/merge.py

```diff
@@ -187,34 +183,121 @@ def handle_merge_conflicts(
+    click.echo(f"\nFound {num_conflicts} conflicting features:")
+    for dataset_path, pks in conflict_pks.items():
+        click.echo(f"{len(pks)} in {dataset_path}")
+    click.echo()
+
+    # Check for dirty working copy before continuing - we don't want to fail after interactive part.
+    ours_rs = repo_structures3[OURS]
+    ours_rs.working_copy.reset(commits3[OURS], ours_rs, force=force)
+
+    # At this point, the failure should be dealt with so we can start resolving conflicts interactively.
+    # We don't want to fail during conflict resolution, since then we would lose all the user's work.
+    # TODO: Support other way(s) of resolving conflicts.
+    if sys.stdout.isatty():
+        _interactive_pause(
+            "Press enter to begin resolving merge conflicts, or Ctrl+C to abort at any time..."
+        )
+    else:
+        click.echo(
+            "Printing conflicts but not resolving - merge conflicts must be resolved in an interactive terminal"
+        )
 
+    # For each conflict, print and maybe resolve it.
     for dataset_path, pks in conflict_pks.items():
         datasets = tuple(rs[dataset_path] for rs in repo_structures3)
+        ours_ds = datasets[OURS]
         for pk in sorted(pks):
+            feature_name = f"{dataset_path}:{pk}"
             features3 = tuple(_safe_get_feature(d, pk) for d in datasets)
-            click.secho(f"=========== {dataset_path}:{pk} ==========", bold=True)
-            for i in range(3):
-                click.secho(f"{big_prefixes3[i]} {labels3[i]}")
-                if features3[i] is not None:
-                    click.secho(
-                        repr_row(features3[i], prefix=prefixes3[i]), fg=cols3[i]
-                    )
-            click.secho()
-
-    raise NotYetImplemented("Sorry, merging conflicts isn't supported yet")
+            print_conflict(feature_name, features3, labels3)
+
+            if sys.stdout.isatty():
+                index_path = os.path.join(dataset_path, ours_ds.get_feature_path(pk))
+                resolve_conflict_interactive(feature_name, merge_index, index_path)
+
+    if not sys.stdout.isatty():
+        raise InvalidOperation(
+            "Merge conflicts must be resolved in an interactive terminal"
+        )
+
+    # Conflicts are resolved, time to commit
+    assert not merge_index.conflicts
+    merge_tree_id = merge_index.write_tree(repo)
+    click.echo(f"Merge tree: {merge_tree_id}")
+
+    user = repo.default_signature
+    merge_message = "Merge '{}'".format(
+        refs3[THEIRS].shorthand if refs3[THEIRS] else refs3[THEIRS].id.hex
+    )
+    commit_id = repo.create_commit(
+        repo.head.name,
+        user,
+        user,
+        merge_message,
+        merge_tree_id,
+        [commits3[OURS].id, commits3[THEIRS].id],
+    )
+    click.echo(f"Merge commit: {commit_id}")
+    return commit_id
+
+
+_version_names3 = ("ancestor", "ours", "theirs")
```

Comment:
could all be `Heads` (🚲🏠-away) namedtuples too 😄

Reply:
Done